File: src/fcgi_writer.c
#include "fcgi_writer.h"
#include "fcgitypes.h"
#include "debug.h"
#include "log.h"

void fcgi_write_buf(struct evbuffer *outBuf, unsigned int requestId, unsigned int type, const char *content, size_t contentLength) {
   const unsigned char paddingLength = (4 - (contentLength % 4)) % 4;
   char padding[] = { 0, 0, 0, 0 };
   char hdr[8];
   hdr[0] = 0x01; // version
   hdr[1] = type;
   hdr[2] = (requestId & 0xff00) > 8;
   hdr[3] = requestId & 0xff;
   hdr[4] = (contentLength & 0xff00) > 8;
   hdr[5] = contentLength & 0xff;
   hdr[6] = paddingLength;
   hdr[7] = 0;    // reserved
   evbuffer_add(outBuf, hdr, sizeof(hdr));
   evbuffer_add(outBuf, content, contentLength);
   evbuffer_add(outBuf, padding, paddingLength);
}

void fcgi_send_response(fd_ctx_t *ctx, const char *response, size_t size) {
   DEBUG("fcgi_send_response");
   struct evbuffer *dst = bufferevent_get_output(ctx->bev);
   fcgi_write_buf(dst, 1, FCGI_STDOUT, response, size);
   fcgi_write_buf(dst, 1, FCGI_STDOUT, "", 0);
   fcgi_write_buf(dst, 1, FCGI_END_REQUEST, "\0\0\0\0\0\0\0\0", 8);
   ctx->closeAfterWrite = true;
}

void fcgi_serve_response(int listen_sock, const char *response, size_t size) {
   struct sockaddr_un client_sockaddr;
   unsigned int len = sizeof(client_sockaddr);

   int client_sock = accept(listen_sock, (struct sockaddr *) &client_sockaddr, &len);
   DEBUG("Accepted client sock: sck%d", client_sock);
   if(client_sock == -1) {
      log_write("[fastcgi spawner] socket accept failed, unable to report an error to std-fpm worker");
      return;
   }

   DEBUG("[fastcgi spawner] socket accepted");

   struct evbuffer *outBuf = evbuffer_new();

   if(!outBuf) {
      log_write("[fastcgi spawner] failed to allocate outBuf for error output");
      return;
   }

   fcgi_write_buf(outBuf, 1, FCGI_STDOUT, response, size);
   fcgi_write_buf(outBuf, 1, FCGI_STDOUT, "", 0);
   fcgi_write_buf(outBuf, 1, FCGI_END_REQUEST, "\0\0\0\0\0\0\0\0", 8);
   evbuffer_write(outBuf, client_sock);

   evbuffer_free(outBuf);
   close(client_sock);
}

File: src/fcgi_writer.rs
use tokio::net::UnixStream;
use tokio::io::{AsyncWriteExt, BufWriter, WriteHalf};


pub struct FcgiWriter {
    wrstream: BufWriter<WriteHalf<UnixStream>>,
}

impl FcgiWriter {
    pub fn new(socket_w: WriteHalf<UnixStream>) -> FcgiWriter {
        FcgiWriter {
            wrstream: BufWriter::new(socket_w),
        }
    }
    pub fn take_socket(self) -> WriteHalf<UnixStream> {
        self.wrstream.into_inner()
    }
    pub async fn send_message(&mut self, msg_type: u8, request_id: u16, content: &[u8]) -> tokio::io::Result<()> {
        let content_length = content.len() as u16;
        let padding_length = ((4 - (content_length % 4)) % 4) as u8;
        let padding = vec![0; 4];
        self.wrstream.write_u8(1).await?; // version
        self.wrstream.write_u8(msg_type).await?;
        self.wrstream.write_u16(request_id).await?;
        self.wrstream.write_u16(content_length).await?;
        self.wrstream.write_u8(padding_length as u8).await?;
        self.wrstream.write_u8(0).await?; // reserved
        self.wrstream.write_all(content).await?;
        self.wrstream.write_all(&padding[0..padding_length as usize]).await?;
        self.wrstream.flush().await?;
        Ok(())
    }
    pub async fn send_response(&mut self, response: &[u8]) -> tokio::io::Result<()> {
        self.send_message(6, 1, response).await?;
        self.send_message(6, 1, &[]).await?;
        Ok(())
    }
}
